dock/core/ui_proxy.h

```diff
@@ -0,0 +1,196 @@
+
+    UIProxy(const std::shared_ptr<ClientNetworkInterface> &networkInterface,
+        bool isTest = false)
+        : mNetworkInterface(networkInterface), mIsTest(isTest) {}
+    
+    static std::shared_ptr<UIProxy> Create(
+        const std::shared_ptr<ClientNetworkInterface> &networkInterface,
+        bool isTest = false) 
+    {
+        return std::make_shared<UIProxy>(networkInterface, isTest);
+    }
+
+    void SwitchToNewStateMachine(
+        const std::shared_ptr<ClientStateMachine> &stateMachine)
+    {
+        assert(mGameManager);
+        mGameManager->SwitchStateMachine(stateMachine);
+    }
+
+    void OnLoginSuccess(const std::string &token, 
+        const std::string &username = "test") 
+    {
+        assert(!mGameManager);
+        mGameManager = std::make_shared<ClientGameManager>(
+            EventLoop::Create(), 
+            SailGame::Dock::StateMachine::Create(), mNetworkInterface);
+        spdlog::info("Game Manager created.");
+        State state;
+        state.mUsername = username;
+        mGameManager->SetState(state);
+
+        std::dynamic_pointer_cast<StateMachine>(mGameManager->GetStateMachine())
+            ->OnGameStart = [this](GameType game) { OnGameStart(game); };
+
+        // token cannot be captured by reference here
+        // because the thread may start running after token gets destructed
+        mGameManagerThread = std::make_unique<std::thread>([token, this] {
+            if (!mIsTest) {
+                mGameManager->StartWithToken(token);
+            }
+            else {
+                CoreMsgBuilder::SetToken(token);
+                // if test, NetworkInterface should Connect instead of AsyncListen
+                mGameManager->Start();
+            }
+        });
+        spdlog::info("Game Manager Thread started.");
+    }
+
+    void Stop() {
+        if (mGameManager) {
+            assert(mGameManagerThread);
+            mGameManager->Stop();
+            mGameManagerThread->join();
+        }
+    }
+    
+    const IState &GetState() const { return mGameManager->GetState(); };
+
+    void SetState(const IState &state) { mGameManager->SetState(state); }
+
+#define RpcMethodHelper(RpcName) \
+    Logger::Log(args); \
+    auto ret = mNetworkInterface->RpcName(args); \
+    Logger::Log(ret); \
+    return ret;
+
+#define RpcMethod(RpcName) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(); \
+    RpcMethodHelper(RpcName);
+
+#define RpcMethod_P1(RpcName, p1) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1); \
+    RpcMethodHelper(RpcName);
+
+#define RpcMethod_P2(RpcName, p1, p2) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1, p2); \
+    RpcMethodHelper(RpcName);
+
+#define RpcMethod_P3(RpcName, p1, p2, p3) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1, p2, p3); \
+    RpcMethodHelper(RpcName);
+
+#define RpcMethod_P4(RpcName, p1, p2, p3, p4) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1, p2, p3, p4); \
```

Comment:
could you try ...args?

Reply:
i'm afraid not. the first argument of the macro is method name, with which it's hard to use that variadic template

Reply:
use variadic macros